crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -283,42 +292,72 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
+                                    .add_negative(rhs_ty)
+                                    .build();
+                                constraints.insert(symbol, ty);
+                            } else {
+                                // Non-singletons cannot be safely narrowed using `is not`
+                            }
                         }
-                    }
-                    ast::CmpOp::Is => {
-                        constraints.insert(symbol, rhs_ty);
-                    }
-                    ast::CmpOp::NotEq => {
-                        if rhs_ty.is_single_valued(self.db) {
-                            let ty = IntersectionBuilder::new(self.db)
-                                .add_negative(rhs_ty)
-                                .build();
-                            constraints.insert(symbol, ty);
+                        ast::CmpOp::Is => {
+                            constraints.insert(symbol, rhs_ty);
+                        }
+                        ast::CmpOp::NotEq => {
+                            if rhs_ty.is_single_valued(self.db) {
+                                let ty = IntersectionBuilder::new(self.db)
+                                    .add_negative(rhs_ty)
+                                    .build();
+                                constraints.insert(symbol, ty);
+                            }
+                        }
+                        _ => {
+                            // TODO other comparison types
                         }
                     }
-                    _ => {
-                        // TODO other comparison types
+                }
+                ast::Expr::Call(ast::ExprCall {
+                    range: _,
+                    func: callable,
+                    arguments,
+                }) if arguments.args.len() == 1 && arguments.keywords.len() == 0 => {
+                    let callable_ty =
+                        inference.expression_ty(callable.scoped_expression_id(self.db, scope));
+
+                    if callable_ty
+                        .into_class_literal()
+                        .is_some_and(|c| c.class.is_known(self.db, KnownClass::Type))
+                    {
+                        if let Some(id) = arguments.args[0].as_name_expr().map(|name| &name.id) {
+                            let symbol = self
+                                .symbols()
+                                .symbol_id_by_name(id)
+                                .expect("Should always have a symbol for every Name node");
+                            if ast::CmpOp::Is == if is_positive { *op } else { op.negate() } {
+                                if rhs_ty.is_class_literal() {
+                                    constraints.insert(symbol, rhs_ty.to_instance(self.db));
+                                }
+                            }
+                        }
                     }
                 }
```

Comment:
What about something like this. It's a bit more verbose, but it moves some cheaper checks higher up (micro-optimising things), improves type safety a little (by using `let` expressions rather than indexing), and reduces the nesting a bit:

```rs
               ast::Expr::Call(ast::ExprCall {
                    range: _,
                    func: callable,
                    arguments:
                        ast::Arguments {
                            args,
                            keywords,
                            range: _,
                        },
                }) => {
                    if !rhs_ty.is_class_literal() {
                        continue;
                    }

                    if !keywords.is_empty() {
                        continue;
                    }

                    let [ast::Expr::Name(ast::ExprName { id, .. })] = &**args else {
                        continue;
                    };

                    let is_valid_constraint = if is_positive {
                        op == &ast::CmpOp::Is
                    } else {
                        op == &ast::CmpOp::IsNot
                    };

                    if !is_valid_constraint {
                        continue;
                    }

                    let callable_ty =
                        inference.expression_ty(callable.scoped_expression_id(self.db, scope));

                    if callable_ty
                        .into_class_literal()
                        .is_some_and(|c| c.class.is_known(self.db, KnownClass::Type))
                    {
                        let symbol = self
                            .symbols()
                            .symbol_id_by_name(id)
                            .expect("Should always have a symbol for every Name node");
                        constraints.insert(symbol, rhs_ty.to_instance(self.db));
                    }
                }
```

Also, could you maybe add a test for the "only a single argument passed to `type()`, but it's a keyword argument" edge case?

Reply:
> Also, could you maybe add a test for the "only a single argument passed to `type()`, but it's a keyword argument" edge case (that an earlier version of this PR panicked on)?

Yes, done.

> What about something like this. It's a bit more verbose, but it moves some cheaper checks higher up (micro-optimising things), improves type safety a little (by using `let ... else` rather than indexing), and reduces the nesting a bit:

I'm not sure. I think I prefer the version with fewer early returns and a more "natural" order of checks. Do you have any evidence that this is faster? Which checks do you consider to be more expensive than others? Even the type-based checks are just lookups of types that have previously been inferred, right?

Reply:
> Do you have any evidence that this is faster? Which checks do you consider to be more expensive than others? Even the type-based checks are just lookups of types that have previously been inferred, right?

no firm evidence, no. My experience working on the Ruff linter, however, has been that checking the structure of the AST is usually much cheaper than looking up cached information in a `HashMap`, which I believe is what both the symbol-table lookup and the `class.is_known()` calls are doing under the hood here.

Still, it might be that for our purposes the difference in performance there is just a rounding error, since red-knot is likely to be significantly slower across the board than a linter that can only do single-file analysis (which is what Ruff currently is). I'm certainly happy to leave it like it is for now and revisit this later if it shows up in profiles.

Reply:
I used your approach and moved a two checks into the match statement